enhancements/ingress/client-tls.md

```diff
@@ -0,0 +1,325 @@
+`ClientCA` field of type `configv1.ConfigMapNameReference` for specifying a
+ConfigMap with a client CA certificate bundle, and an optional
+`AllowedSubjectPatterns` field of type `[]string` for configuring a list of
+allowed subject names:
+
+```go
+// ClientTLS specifies TLS configuration to enable client-to-server
+// authentication, which can be used for mutual TLS.
+type ClientTLS struct {
+	// clientCertificatePolicy specifies whether the ingress controller
+	// requires clients to provide certificates.  This field accepts the
+	// values "Required" or "Optional".
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	ClientCertificatePolicy ClientCertificatePolicy `json:"clientCertificatePolicy"`
+
+	// clientCA is an optional reference to a configmap containing the
+	// PEM-encoded CA certificate bundle that should be used to verify a
+	// client's certificate.  If omitted, the system trust bundle is used.
+	//
+	// +optional
+	ClientCA configv1.ConfigMapNameReference `json:"clientCA"`
+
+	// allowedSubjectPatterns specifies a list of regular expressions that
+	// should be matched against the distinguished name on a client
+	// certificate to filter requests.  If this list is empty, no filtering
+	// is performed.  If the list is nonempty, then at least one pattern
+	// must match a client certificate's distinguished name or else the
+	// ingress controller rejects the certificate and denies the connection.
+	//
+	// +optional
+	AllowedSubjectPatterns []string `json:"allowedSubjectPatterns,omitempty"`
+}
+```
+
+The `ClientCertificatePolicy` type accepts either one of two values: `Required`
+or `Optional`:
+
+```go
+// ClientCertificatePolicy describes the policy for client certificates.
+// +kubebuilder:validation:Enum="";Required;Optional
+type ClientCertificatePolicy string
+
+const (
+	// ClientCertificatePolicyRequired indicates that a client certificate
+	// should be required.
+	ClientCertificatePolicyRequired ClientCertificatePolicy = "Required"
+
+	// ClientCertificatePolicyOptional indicates that a client certificate
+	// should be requested but not required.
+	ClientCertificatePolicyOptional ClientCertificatePolicy = "Optional"
+)
+```
+
+If the user does not specify `spec.clientTLS`, then client TLS is not enabled,
+which means that the IngressController does not request client certificates on
+TLS connections.  If `spec.clientTLS` is specified, then the IngressController
+does request client certificates, and `spec.clientTLS.clientCertificatePolicy`
+must be specified to indicate whether the IngressController should reject
+clients that do not provide valid certificates.
+
+The optional `ClientCA` field may be used to specify a reference to a configmap
+that is in the same namespace as the IngressController and contains a CA
+certificate bundle.  If `ClientCA` is omitted, then the IngressController uses
+the system trust bundle.  If `ClientCA` is specified, then the IngressController
```

Comment:
I think the use of the system trust bundle should be turned on with a flag, not the default. There is an inherent risk of letting in certs that you do not intend to when you use the system trust.

Reply:
What should the behavior be when `clientCA` is omitted and the flag to use the system trust bundle is not set?  If both `clientCA` and the flag are set, do we need to combine bundles?  

The more I think about this, the more I think we should just skip the option to use the system trust bundle, and instead require the user to specify a bundle using an API object (such as a ConfigMap or PVC).

Reply:
I would not use the system trust bundle to terminate client certificates.  In kube, we don't even have a way to express this.